lib/script-generator.js

```diff
@@ -0,0 +1,209 @@
+  };
+}
+
+/**
+ * Sets the rules for the script generator.
+ * @param {array} rules Rules to set for the generator.
+ */
+ScriptGenerator.prototype.setRules = function (rules) {
+  this.rules = rules;
+};
+
+/**
+ * Generates a shell script from the rules.
+ * @return {string} A shell script generated from the provided rules.
+ */
+ScriptGenerator.prototype.generate = function() {
+  return [
+    this.preamble(),
+    this.rules.map(this.generateRule.bind(this)).join('\n')
+  ].join('\n');
+};
+
+/**
+ * Generates the preamble for the shell script.
+ * @return {string} The preamble for the script.
+ */
+ScriptGenerator.prototype.preamble = function () {
+  var header = [
+    '#!/bin/sh\n',
+    '#',
+    '# Warning: this is a generated file, modifications may be overwritten.',
+    '#\n'
+  ].join('\n');
+
+  var common = [
+    '_script_name=`basename $0`',
+    'search_files=\'.\'',
+    'warning() { echo "($_script_name) WARNING" $1; }',
+    'error() { echo "($_script_name) ERROR" $1; exit 1; }\n'
+  ].join('\n');
+
+  var commandChecks = [
+    FsDriver.commands.map(function (name) {
+      return 'command -v ' + name + ' >/dev/null 2>&1 || {\n' +
+        '  error "Missing required command: ' + name + '";\n' +
+        '}';
+    }).join('\n'),
+  ].join('\n') + '\n';
+
+  return [header, common, commandChecks].join('\n');
+};
+
+/**
+ * Generates the script for the given rule.
+ * @param {object} rule Rule for which to generate the script.
+ * @return {string} The script that handles the given rule.
+ */
+ScriptGenerator.prototype.generateRule = function (rule, index) {
+  return this.actionHandlers[rule.action](rule, index);
+};
+
+/**
+ * Generates the script for a copy rule.
+ * @param {object} rule Copy rule.
+ * @return {string} Script for the given rule.
+ */
+ScriptGenerator.prototype.copy = function (rule, index) {
+  var header = [
+    '# RULE ' + index,
+    '# {',
+    '#   action: "' + rule.action + '",',
+    '#   source: "' + rule.source + '",',
+    '#   dest: "' + rule.dest + '"',
+    '# }'
+  ].join('\n') + '\n';
+
+  var body = [
+    'cp ' + rule.source + ' ' + rule.dest + ' || {',
+    '  warning "Rule ' + index + ': unable to copy ' +
+      rule.source + ' to ' + rule.dest + '"',
+    '}'
+  ].join('\n');
+
+  return [header, body].join('\n') + '\n';
+};
+
+/**
+ * Generates the script for a rename rule.
+ * @param {object} rule Rename rule.
+ * @return {string} Script for the given rule.
+ */
+ScriptGenerator.prototype.rename = function (rule, index) {
+  var header = [
+    '# RULE ' + index,
+    '# {',
+    '#   action: "' + rule.action + '",',
+    '#   source: "' + rule.source + '",',
+    '#   dest: "' + rule.dest + '"',
+    '# }'
+  ].join('\n') + '\n';
+
+  var body = [
+    'mv ' + rule.source + ' ' + rule.dest + ' || {',
+    '  warning "Rule ' + index + ': unable to rename ' +
+      rule.source + ' to ' + rule.dest + '"',
+    '}'
+  ].join('\n');
+
+  return [header, body].join('\n') + '\n';
+};
```

Comment:
in these generating blocks, why not just add an empty string as the last item in the array - then you avoid having to remember these `+ \n` every once in a while...?